accord-core/src/main/java/accord/utils/SortedArrays.java

```diff
@@ -447,6 +447,93 @@ else if (cmp < 0)
         }
     }
 
+    // applies the update function to all entries in left, and any matching entries in right
+    public static <O, T extends O> O[] linearUpdate(T[] left, int leftStart, int leftEnd, T[] right, int rightStart, int rightEnd, AsymmetricComparator<? super T, ? super T> comparator, BiFunction<? super T, ? super T, ? extends T> update, ObjectBuffers<O> buffers)
+    {
+        int leftIdx = leftStart;
+        int rightIdx = rightStart;
+
+        T nextKey = null;
+        while (leftIdx < leftEnd && rightIdx < rightEnd)
+        {
+            T leftKey = left[leftIdx];
+            T rightKey = right[rightIdx];
+            int cmp = leftKey == rightKey ? 0 : comparator.compare(leftKey, rightKey);
+            if (cmp > 0)
+            {
+                ++rightIdx;
+                continue;
+            }
+
+            if (cmp < 0) rightKey = null;
+            else ++rightIdx;
+
+            T testKey = update.apply(leftKey, rightKey);
+            if (testKey != leftKey)
+            {
+                nextKey = testKey;
+                break;
+            }
+            ++leftIdx;
+        }
+
+        if (nextKey == null)
+        {
+            while (leftIdx < leftEnd)
+            {
+                T leftKey = left[leftIdx];
+                T testKey = update.apply(left[leftIdx], null);
+                if (testKey != leftKey)
+                {
+                    nextKey = testKey;
+                    break;
+                }
+                ++leftIdx;
+            }
+
+            if (nextKey == null)
+                return left;
+        }
+
+        O[] result = buffers.get(leftEnd - leftStart);
+        int resultSize = leftIdx - leftStart;
+        System.arraycopy(left, leftStart, result, 0, resultSize);
+        result[resultSize++] = nextKey;
+        ++leftIdx;
+
+        try
+        {
+            while (leftIdx < leftEnd && rightIdx < rightEnd)
+            {
+                T leftKey = left[leftIdx];
+                T rightKey = right[rightIdx];
+                int cmp = leftKey == rightKey ? 0 : comparator.compare(leftKey, rightKey);
+
+                if (cmp > 0)
+                {
+                    ++rightIdx;
+                    continue;
+                }
+
+                if (cmp < 0) rightKey = null;
+                else ++rightIdx;
+                ++leftIdx;
+
+                result[resultSize++] = update.apply(leftKey, rightKey);
+            }
+
+            while (leftIdx < leftEnd)
```

Comment:
Just to make sure: do we always know that left is longer than right?

Reply:
we don't, we just don't care if right is longer than left - this is meant to be like a left join; we only care about those items on the right that match those on the left.